### PR TITLE
add increment development function

### PR DESCRIPTION
**Pull request by mannaandpoem**

### CODE REVIEW

**Comment by stellaHSR** [comment]
Good job

**Comment by geekan** [comment]
Wait for you to merge the logic into the latest code

**Comment by mannaandpoem** [comment]
> logic

The simplest logic has been merged into PR. Next, I need to fine tune the prompt of WritePRD, WriteDesign and WriteTasks to improve the effect.

> Wait for you to merge the logic into the latest code

The simplest logic has been merged into PR. Next, I need to fine tune the prompt of WritePRD, WriteDesign and WriteTasks to improve the effect.

**Comment by mannaandpoem** [comment]
> Wait for you to merge the logic into the latest code

The simplest logic has been merged into PR. Next, we need to fine-tune the prompt words of WritePRD, WriteDesign and WriteTasks to improve the effect.

**Comment by geekan** [comment]
At the conclusion of the meeting, we need to increase the process of minor repairs

File: metagpt/actions/write_prd_an.py
Comment (geekan) [inline metagpt/actions/write_prd_an.py:228]: lacks of incremental

**Review by geekan — CHANGES_REQUESTED** [review]
The overall implementation is as expected, but the style needs to be complete and consistent, or the style needs to be appropriately annotated.
For example, the key naming convention corresponding to the action node of incremental/refine. These need to be modified before being merged

and related test cases need to be completed. But this can be done again when merge dev branch

File: metagpt/actions/project_management_an.py
Comment (mannaandpoem) [reply metagpt/actions/project_management_an.py:49]: OK, I originally wanted to make them consistent with the original PRD or Design or Task for subsequent parse

File: tests/metagpt/test_increment.py
Comment (geekan) [inline tests/metagpt/test_increment.py:149]: The files provided here are not accompanied by uploads. Perhaps a best practice is to put it in the data/ folder instead of using DEFAULT_WORKSPACE_ROOT

**Review by geekan — CHANGES_REQUESTED** [review]

File: tests/metagpt/test_increment.py
Comment (geekan) [inline tests/metagpt/test_increment.py:22]: 1. I mean a const path "DATA_PATH".
2. The packages such as "simple_add_calculator" should be uploaded as zip / tar.gz with related unzip / untar commands.
3. test_increment.py is somewhat unclear, is test_incremental_dev.py better?

**Review by geekan — CHANGES_REQUESTED** [review]

File: tests/metagpt/test_increment.py
Comment (mannaandpoem) [reply tests/metagpt/test_increment.py:22]: OK, I will continue to modify the test cases.

Comment (mannaandpoem) [reply tests/metagpt/test_increment.py]: Modified the key code and uploaded 4 code compression packages (maybe not appropriate, better ones need to be given)

File: metagpt/roles/engineer.py
Comment (stellaHSR) [inline metagpt/roles/engineer.py:346]: Add function comments

File: metagpt/actions/write_code.py
Comment (stellaHSR) [inline metagpt/actions/write_code.py:159]: Please add comments to the function, specifically for the 'mode' parameter, to clarify its purpose.

File: metagpt/actions/project_management_an.py
Comment (better629) [inline metagpt/actions/project_management_an.py:129]: These INC/MERGE_XX_CONTEXT not used?

File: metagpt/actions/write_code.py
Comment (better629) [inline metagpt/actions/write_code.py:176]: why exclude `main.py` ?

File: metagpt/roles/engineer.py
Comment (better629) [inline metagpt/roles/engineer.py:122]: simplify code
```
dependencies = {
                    coding_context.design_doc.root_relative_path,
                    coding_context.task_doc.root_relative_path,
                }
if guideline:
    dependencies.add(xx)

```

File: tests/metagpt/actions/test_design_api_an.py
Comment (better629) [inline tests/metagpt/actions/test_design_api_an.py:96]: seems no need to add this

File: tests/metagpt/actions/test_design_api_an.py
Comment (better629) [inline tests/metagpt/actions/test_design_api_an.py:104]: suggest to check the complete key in `x.model_dump` not just the content.

File: tests/metagpt/test_incremental_dev.py
Comment (better629) [inline tests/metagpt/test_incremental_dev.py:22]: if the files under MetaGPT/data is only for unittest, suggest move to MetaGPT/tests/data

File: tests/metagpt/test_incremental_dev.py
Comment (better629) [inline tests/metagpt/test_incremental_dev.py:25]: How many type of increment idea support? like add new feature/ change existed feature and so on. Can we list these and add them in the unittests?

And seems that the given args[0] including increment idea and its current situation, does the current situation need ? seems that it will provide with more context.

File: tests/metagpt/test_incremental_dev.py
Comment (better629) [inline tests/metagpt/test_incremental_dev.py:62]: Can we add a base function to only pass project_path and increment context to simplify the code. Most of them are repetitive.

File: metagpt/actions/write_code_guideline_an.py
Comment (better629) [inline metagpt/actions/write_code_guideline_an.py:458]: use logger.info

File: metagpt/actions/design_api_an.py
Comment (better629) [inline metagpt/actions/design_api_an.py:94]: This is a discussion, it seems that all increment or refined should refine the ActionNode, But most of the instruction are similar. So, Can we add a IncreActionNode/ RefineActionNode to follow a incre/refine **format** to extend the instruction, seems that will help the user to only handle the original one(PROGRAM_CALL_FLOW) but not with three(PROGRAM_CALL_FLOW INCRE_XX REFINED_XX).

**Comment by better629** [comment]
1. Due to this isn't a small pr, add the unittest result to help reviewer to check.
2. Add some description to explain how to add incre/refine due to only some nodes need these two operations. And the children nodes to be modified are tightly coupled to the process, how can we help the user to understand these detail changes.

File: metagpt/actions/design_api.py
Comment (better629) [reply metagpt/actions/design_api.py:112]: refs to https://github.com/better629/MetaGPT/commit/1238c484511e4a54691fe2816b43833eac41399d .

Comment (better629) [reply metagpt/actions/design_api.py:112]: refs to https://github.com/better629/MetaGPT/commit/1238c484511e4a54691fe2816b43833eac41399d .

Comment (mannaandpoem) [reply metagpt/actions/design_api.py:112]: Can communicate with DATA_ Structures_ AND_ Maintain consistency with INTERFACES, but this will result in inconsistent variable names with ActionNode

File: metagpt/actions/design_api_an.py
Comment (mannaandpoem) [reply metagpt/actions/design_api_an.py:76]: Currently, all INC_XXX is not in use, I will delete it soon

File: metagpt/actions/project_management_an.py
Comment (mannaandpoem) [reply metagpt/actions/project_management_an.py:129]: They are not in use, I am about to delete them

File: metagpt/actions/project_management_an.py
Comment (mannaandpoem) [reply metagpt/actions/project_management_an.py:43]: 好的👌

File: metagpt/actions/write_code.py
Comment (mannaandpoem) [reply metagpt/actions/write_code.py:176]: In most cases, main.py does not contain too much code, and it is not necessary to refer to legacy main.py to generate good code. The existence of legacy main.py may lead to the generation of a lot of unnecessary content.

Comment (better629) [reply metagpt/actions/write_code.py:176]: Usually, the llm will generate code in one file for small projects. It's our requirement that split the `main.py` and `other_sourcecode.py`. May here should add an explanation about excluding.

File: metagpt/actions/design_api_an.py
Comment (geekan) [inline metagpt/actions/design_api_an.py:46]: up to 2 files, but here are 3

File: metagpt/actions/project_management_an.py
Comment (geekan) [inline metagpt/actions/project_management_an.py:121]: It is recommended to use all capitalization

File: metagpt/actions/write_code.py
Comment (geekan) [inline metagpt/actions/write_code.py:120]: a = ...
b = func(mode=a)

File: metagpt/actions/write_code.py
Comment (geekan) [inline metagpt/actions/write_code.py:160]: mode: Literal[...]

File: metagpt/actions/write_code.py
Comment (geekan) [inline metagpt/actions/write_code.py:194]: When should you specify exclude?

File: metagpt/actions/write_code.py
Comment (geekan) [inline metagpt/actions/write_code.py:205]: More comments are needed here, and the design logic is not discernible from the code alone

File: metagpt/actions/write_code.py
Comment (geekan) [inline metagpt/actions/write_code.py:211]: If the above is a Literal, the corresponding condition should be explicitly used here. Even if it's not explicitly used, it should be annotated

File: metagpt/actions/write_code.py
Comment (geekan) [inline metagpt/actions/write_code.py:166]: why should we use exclude?

**Comment by stellaHSR** [comment]
LGTM

File: metagpt/actions/write_code_guideline_an.py
Comment (geekan) [inline metagpt/actions/write_code_guideline_an.py:14]: Guideline is a not-so-good name. Plan is a bit better than that

File: metagpt/actions/write_code_review.py
Comment (geekan) [inline metagpt/actions/write_code_review.py:142]: dont hard code this.

File: metagpt/roles/engineer.py
Comment (geekan) [inline metagpt/roles/engineer.py:103]: Keep it in the same style as .key before

File: metagpt/roles/engineer.py
Comment (geekan) [inline metagpt/roles/engineer.py:347]: Keep the style consistent

File: metagpt/utils/mermaid.py
Comment (geekan) [inline metagpt/utils/mermaid.py:136]: In Mermaid, this form of annotation is illegal. This will cause the output to not be rendered

File: metagpt/utils/mermaid.py
Comment (geekan) [inline metagpt/utils/mermaid.py:155]: ditto.

File: metagpt/utils/mermaid.py
Comment (geekan) [inline metagpt/utils/mermaid.py:210]: ditto

File: tests/metagpt/test_incremental_dev.py
Comment (geekan) [inline tests/metagpt/test_incremental_dev.py:108]: project_path.exists()

File: metagpt/actions/project_management_an.py
Comment (better629) [inline metagpt/actions/project_management_an.py:121]: from_children("REFINED_PM_NODE")

File: metagpt/actions/write_code.py
Comment (better629) [inline metagpt/actions/write_code.py:178]: why name it as "plan"? Yizhang is working wth "Plan", will the name make a confuse?

File: metagpt/actions/write_code_plan_an.py
Comment (better629) [inline metagpt/actions/write_code_plan_an.py:12]: Can the "Plan" to name with more clear one?

File: metagpt/actions/write_code_plan_an.py
Comment (better629) [inline metagpt/actions/write_code_plan_an.py:184]: Change.

File: metagpt/actions/write_prd.py
Comment (better629) [inline metagpt/actions/write_prd.py:138]: `REFINE_PRD_TEMPLATE` keep same with `REFINED_CODE_TEMPLATE` ? use `REFINED`, check all the context.

File: metagpt/actions/write_prd_an.py
Comment (better629) [inline metagpt/actions/write_prd_an.py:189]: REFINED

File: metagpt/actions/write_prd_an.py
Comment (better629) [inline metagpt/actions/write_prd_an.py:219]: REFINED

File: tests/metagpt/actions/test_write_prd_an.py
Comment (better629) [inline tests/metagpt/actions/test_write_prd_an.py:35]: REFINED_PRD_NODE

**Comment by geekan** [comment]
Merge first. Follow-up optimization

**Review by geekan — APPROVED** [review]
LGTM

**Comment by better629** [comment]
LGTM

**Pull request merged**
